File: neural_network/main/trainer.py
import math
import pandas as pd

from neural_network.components import Network

from .plotter import Plotter
from .abstract_simulator import AbstractSimulator
from .validator import Validator


class Trainer(AbstractSimulator):
    """Class to train a neural network
    """

    def __init__(self, network: Network, data: pd.DataFrame, num_epochs: int,
                 batch_size: int, validator: Validator = None,
                 weighted: bool = False, classification: bool = True):
        """Constructor method

        Parameters
        ----------
        network : Network
            The neural network to train
        data : pd.DataFrame
            All the training data for the `Network`
        num_epochs : int
            The number of epochs we are training for
        batch_size : int
            The number of datapoints used in each epoch
        validator : Validator
            The validator used (if any)
        weighted : bool
            If `True` then we use the WeightedPartitioner, otherwise we use
            the standard Partitioner
        classification : bool
            If `True` then we are classifying, otherwise it will be regression
        """
        super().__init__(network, data, batch_size, weighted, classification)
        self._num_epochs = num_epochs
        self._validator = validator

        columns = ['Training']
        if self._validator:
            columns.append('Validation')
        self._loss_df = pd.DataFrame(columns=columns)

    def store_gradients(self, _id: int):
        """Stores the gradients of the loss functions after a forward pass

        Parameters
        ----------
        _id : int
            The id of the datapoint
        """
        y = int(self._data.loc[_id, 'y'])

        # Take gradients of loss and store them in the edges (backwards)
        edges = self._network.get_edges()
        for edge_layer in reversed(edges):
            for right_neuron in edge_layer:
                first = True
                for edge in right_neuron:
                    self._network.store_gradient_of_loss(edge, y, first)
                    first = False

    def back_propagate_one_batch(self):
        """Performs back propagation for one batch of datapoints (stored within
        the memory of the edges).
        """
        edges = self._network.get_edges()
        for layer in reversed(edges):
            for right_neuron in layer:
                for edge in right_neuron:
                    self._network.back_propagate_weight(edge)

        layers = self._network.get_layers()
        for layer in layers[1:]:
            for neuron in layer.get_neurons():
                self._network.back_propagate_bias(neuron)

    def run(self):
        """Performs training of the network
        """
        factor = math.ceil(self._num_epochs / 100)
        for epoch in range(self._num_epochs):
            total_loss = 0
            batch_partition = self._partitioner()
            for iteration in range(math.ceil(len(self._data)
                                             / self._batch_size)):
                batch_ids = batch_partition[iteration]
                total_loss += self.forward_pass_one_batch(batch_ids)
                self.back_propagate_one_batch()
            loss = round(total_loss / len(self._data), 8)
            if epoch % factor == 0:
                print(f"Epoch: {epoch}")
                print(f"Loss: {loss}")

            self._loss_df.at[epoch, 'Training'] = loss
            if self._validator:
                validation_loss = self._validator.validate(factor)
                self._loss_df.at[epoch, 'Validation'] = validation_loss

        if not self._regression:
            self._update_categorical_dataframe()

    def generate_scatter(self, title: str = ''):
        """Creates scatter plot from the data and their predicted values

        Parameters
        ----------
        title : str
            An optional title to append to the plot
        """
        super().abs_generate_scatter(phase='training', title=title)

    def generate_loss_plot(self, title: str = ''):
        Plotter.plot_loss(self._loss_df, title)

File: javaNeuralNetwork/src/main/java/neural_network/learning/Trainer.java
package neural_network.learning;

import neural_network.components.Edge;
import neural_network.components.Network;
import neural_network.util.Header;
import neural_network.util.Plotter;

import java.io.IOException;
import java.util.*;

/** Class to train a neural network.
 *
 */
public class Trainer extends Learner {

    private final int numEpochs;
    private Validator validator;
    private final Map<String, List<Double>> lossDf = new TreeMap<>();

    /** General constructor method.
     *
     * @param network The neural network to train.
     * @param data All the training data for the {@code network}.
     * @param batchSize The number of datapoints used in each epoch.
     * @param weighted If {@code true} then we use the {@code WeightedPartitioner},
     *                 otherwise we use the standard {@code Partitioner}.
     * @param numBins If {@code weighted} is {@code true} and the {@code network} is
     *                a regressor, then we need to specify the number of bins for the
     *                {@code WeightedPartitioner}. Otherwise, this parameter is ignored.
     * @param numEpochs The number of epochs we are training for.
     * @param validator The validator used (if any). Pass {@code null}
     *                  for no validation.
     */
    public Trainer(Network network, NavigableMap<Header, List<String>> data,
                   int batchSize, boolean weighted, int numBins, int numEpochs,
                   Validator validator) {
        super(network, data, batchSize, weighted, numBins);
        this.numEpochs = numEpochs;
        this.validator = validator;
        this.lossDf.put("Training", new ArrayList<>());
        if (validator != null) {
            this.lossDf.put("Validation", new ArrayList<>());
        }
    }

    /** Defaults constructor. Used for unweighted partitioner and no validation.
     *
     * @param network The neural network to train.
     * @param data All the training data for the {@code network}.
     * @param batchSize The number of datapoints used in each epoch.
     * @param numEpochs The number of epochs we are training for.
     */
    public Trainer(Network network, NavigableMap<Header, List<String>> data,
                   int batchSize, int numEpochs) {
        super(network, data, batchSize);
        this.numEpochs = numEpochs;
        this.lossDf.put("Training", new ArrayList<>());
    }

    /** Stores the gradients of the loss functions after a forward pass of
     * a single datapoint.
     *
     * @param id The id of the current datapoint.
     */
    @Override
    void storeGradients(int id) {
        double y = getDf().get(Header.Y).get(id);

        // Take gradients of loss and store them in the edges (backwards)
        List<List<List<Edge>>> edges = getNetwork().getEdges();
        Collections.reverse(edges);
        for (List<List<Edge>> edgeLayer : edges) {
            for (List<Edge> rightNeuron : edgeLayer) {
                boolean first = true;
                for (Edge edge : rightNeuron) {
                    getNetwork().storeGradientOfLoss(edge, y, first);
                    first = false;
                }
            }
        }
    }

    /** Performs back propagation for one batch of datapoints (stored within
     * the memory of the edges).
     *
     */
    void backPropagateOneBatch() {
        // Back propagate all the weights first
        getNetwork().backPropagateWeights();

        // Then back propagate all the biases
        getNetwork().backPropagateBiases();
    }

    /** Performs training of the network.
     *
     */
    @Override
    public void run() {
        int factor = (int) Math.ceil((double) numEpochs / 100);
        for (int epoch = 0; epoch < numEpochs; epoch ++) {
            double totalLoss = 0.0;
            // Partition all the datapoints into batches
            List<List<Integer>> batchPartition = getPartitioner().call();
            int itsPerEpoch = (int) Math
                    .ceil((double) getNumDatapoints() / getBatchSize());
            for (int iteration = 0; iteration < itsPerEpoch; iteration ++) {
                List<Integer> batchIds = batchPartition.get(iteration);

                // Do forward pass and back propagation for this specific batch
                totalLoss += forwardPassOneBatch(batchIds);
                backPropagateOneBatch();
            }
            double loss = Math.round(10000 * totalLoss / getNumDatapoints()) / 10000.0;
            if (epoch % factor == 0) {
                System.out.println("Epoch: " + epoch);
                System.out.printf("Training loss: %.4f%n", loss);
            }
            // Record the loss and potential validation loss
            lossDf.get("Training").add(loss);
            if (validator != null) {
                double validationLoss = validator.validate(factor);
                lossDf.get("Validation").add(validationLoss);
            }
        }
        // At the end, update the categorical dataframe
        if (! isRegressor()) {
            updateCategoricalDataframe();
        }
    }

    /** Creates scatter plot from the data and their predicted values.
     *
     * @param title An optional title to append to the plot.
     * @throws IOException If an IO error occurs.
     */
    public void generateScatter(String title) throws IOException {
        super.generateScatter("training", title);
    }

    /** Creates scatter plot comparing predicted to actual values (for
     * regression problems only).
     *
     * @param title An optional title to append to the plot.
     * @throws IOException If an IO error occurs.
     * @throws RuntimeException If this method is called with a categorical network,
     *                          instead user should call {@code Tester.generateConfusion()}.
     */
    public void comparisonScatter(String title) throws IOException {
        super.comparisonScatter("training", title);
    }

    /** Creates a plot of the training and validation loss over time.
     *
     * @param title An optional title to append to the plot.
     * @throws IOException If an IO error occurs.
     */
    public void generateLossPlot(String title) throws IOException {
        Plotter.plotLoss(lossDf, title);
    }

    /** Getter for {@code lossDf}.
     *
     * @return A deep copy of the {@code lossDf}.
     */
    public Map<String, List<Double>> getLossDf() {
        Map<String, List<Double>> returnMap = new HashMap<>();
        for (String header : lossDf.keySet()) {
            returnMap.put(header, List.copyOf(lossDf.get(header)));
        }
        return returnMap;
    }
}
